# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/vm/JSAtomUtils.cpp
# Commit: fa120d93961e
# Full Hash: fa120d93961efddd63d50135f03c55e75dac875a
# Author: Doug Thayer <dothayer@mozilla.com>
# Date: 2024-04-26 04:18:06
# Regressor Bug: 1881995
# File Overlap Count: 1
# Description:
#   Bug 1881995 - Implement ForwardedAtoms and create them during atomization r=iain,sfink
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D202690
# ==============================================================================

diff -r f24ffc487097 -r fa120d93961e js/src/vm/JSAtomUtils.cpp
--- a/js/src/vm/JSAtomUtils.cpp	Thu Apr 25 21:30:10 2024 +0300
+++ b/js/src/vm/JSAtomUtils.cpp	Thu Apr 25 18:29:15 2024 +0000
@@ -666,6 +666,10 @@
     return &str->asAtom();
   }
 
+  if (str->isAtomRef()) {
+    return str->atom();
+  }
+
   if (JSAtom* atom = cx->caches().stringToAtomCache.lookup(str)) {
     return atom;
   }
@@ -691,6 +695,7 @@
         // not done in lookup() itself, because #including JSContext.h there
         // causes some non-trivial #include ordering issues.
         cx->markAtom(atom);
+        str->tryReplaceWithAtomRef(atom);
         return atom;
       }
     }
@@ -723,7 +728,9 @@
     return nullptr;
   }
 
-  cx->caches().stringToAtomCache.maybePut(str, atom, key);
+  if (!str->tryReplaceWithAtomRef(atom)) {
+    cx->caches().stringToAtomCache.maybePut(str, atom, key);
+  }
 
   return atom;
 }